libs/assertion/include/hpx/assertion.hpp

```diff
@@ -58,16 +58,27 @@ namespace hpx { namespace assertion {
             __FILE__, __LINE__, HPX_ASSERT_CURRENT_FUNCTION},                  \
         HPX_PP_STRINGIZE(expr), [&]() { return msg; });                        \
 /**/
+#define HPX_ASSERT_EQ(x, count)
```

Comment:
This definition is not necessary (and leads to errors due to `HPX_ASSERT_EQ` begin redefined). I'd also suggest naming the two arguments e.g. `expr1` and `expr2` to be consistent with the `HPX_TEST_*` macros and since the second argument isn't necessarily a number.

Reply:
What?! How do you visualise the `HPX_ASSERT_EQ(x, 3);` item with `HPX_ASSERT_EQ(expr1, expr2);`? 

- Do you require a macro equating two expressions? and returning a boolean value?
- Do you require a macro equating x and a digit and return digit if equating fails?

Reply:
Please have a look at https://github.com/STEllAR-GROUP/hpx/blob/master/libs/testing/include/hpx/testing.hpp to get some ideas on how you can implement it.

Reply:
Can you please answer:

- Do you require a macro equating two expressions? and returning a boolean value?
- Do you require a macro equating expression and a value and return digit if equating fails?

Ref: https://github.com/STEllAR-GROUP/hpx/issues/3816#issuecomment-527876871.

Reply:
The `HPX_ASSERT_EQ` macro should take two *expressions*, with the only requirement that the types of the two expressions are equatable (have operator `==`) and that they can be written to a stream (have operator `<<`). A macro in general can not return a value, but the macro should also not evaluate to a value. The macro should check for equality between the two values and if they are not equal call `handle_assert` with a message which prints both the expression (`expr1 == expr2`) and the evaluated values of the two expressions (e.g. `1 == 2`, or `"abc" == "def"`). This is where the current `HPX_ASSERT` macro eventually ends up: https://github.com/STEllAR-GROUP/hpx/blob/414380e50e55ed1f4ebfde57f3bda7018d6d1cf0/libs/assertion/include/hpx/assertion/evaluate_assert.hpp#L22-L30, and then: https://github.com/STEllAR-GROUP/hpx/blob/414380e50e55ed1f4ebfde57f3bda7018d6d1cf0/libs/assertion/src/assertion.cpp#L31-L48. It may be easiest to add functions `evaluate_assert_eq`, `evaluate_assert_neq` etc. functions to handle the different comparisons. Does that make it clearer?

Reply:
Cool, ok.